packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol

```diff
@@ -241,6 +241,32 @@ contract L1GatewayRouter is WhitelistConsumer, L1ArbitrumMessenger, GatewayRoute
         return super.outboundTransfer(_token, _to, _amount, _maxGas, _gasPriceBid, _data);
     }
 
+    function outboundTransferCustomRefund(
+        address _token,
+        address _refundTo,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data
+    ) public payable override onlyWhitelisted returns (bytes memory) {
+        // _refundTo is subject to L2 alias rewrite
+        require(_refundTo != address(0), "INVALID_REFUND_ADDR");
+        // when sending a L1 to L2 transaction, we expect the user to send
+        // eth in flight in order to pay for L2 gas costs
+        // this check prevents users from misconfiguring the msg.value
+        (uint256 _maxSubmissionCost, ) = abi.decode(_data, (uint256, bytes));
+
+        // here we don't use SafeMath since this validation is to prevent users
+        // from shooting themselves on the foot.
+        uint256 expectedEth = _maxSubmissionCost + (_maxGas * _gasPriceBid);
+        require(_maxSubmissionCost > 0, "NO_SUBMISSION_COST");
+        require(msg.value == expectedEth, "WRONG_ETH_VALUE");
```

Comment:
As opposed to the changes in #2128, I think maybe refactoring this chunk into its own function might make sense since the logic is relatively complex. It's significantly more likely we're going to want to mess with this logic later than the stuff from GatewayRouter.sol, eg. for nitro